src/server/tablet_snap.rs

```diff
@@ -0,0 +1,537 @@
+        }
+        let meta = head.take_message();
+        let snapshot = meta.get_message().get_snapshot();
+        let key = TabletSnapKey::from_region_snap(
+            meta.get_region_id(),
+            meta.get_to_peer().get_id(),
+            meta.get_message().get_snapshot(),
+        );
+
+        let data = snapshot.get_data();
+        let mut snapshot_data = RaftSnapshotData::default();
+        snapshot_data.merge_from_bytes(data)?;
+        let snapshot_meta = snapshot_data.take_meta();
+        let io_type = if snapshot_meta.get_for_balance() {
+            IoType::LoadBalance
+        } else {
+            IoType::Replication
+        };
+
+        Ok(RecvTabletSnapContext {
+            key,
+            raft_msg: meta,
+            io_type,
+            start: Instant::now(),
+        })
+    }
+
+    fn finish<R: RaftStoreRouter<impl KvEngine>>(self, raft_router: R) -> Result<()> {
+        let _with_io_type = WithIoType::new(self.io_type);
+        let key = self.key;
+        if let Err(e) = raft_router.send_raft_msg(self.raft_msg) {
+            return Err(box_err!("{} failed to send snapshot to raft: {}", key, e));
+        }
+        info!("saving all snapshot files"; "snap_key" => %key, "takes" => ?self.start.saturating_elapsed());
+        Ok(())
+    }
+}
+
+async fn send_snap_files(
+    mgr: &TabletSnapManager,
+    mut sender: impl Sink<(SnapshotChunk, WriteFlags), Error = Error> + Unpin,
+    msg: RaftMessage,
+    key: TabletSnapKey,
+    limiter: Limiter,
+) -> Result<u64> {
+    let path = mgr.get_final_path_for_gen(&key);
+    info!("begin to send snapshot file";"snap_key" => %key);
+    let files = fs::read_dir(&path)?
+        .map(|f| Ok(f?.path()))
+        .filter(|f| f.is_ok() && f.as_ref().unwrap().is_file())
+        .collect::<Result<Vec<_>>>()?;
+    let mut total_sent = msg.compute_size() as u64;
+    let mut chunk = SnapshotChunk::default();
+    chunk.set_message(msg);
+    sender
+        .feed((chunk, WriteFlags::default().buffer_hint(true)))
+        .await?;
+    for path in files {
+        let name = path.file_name().unwrap().to_str().unwrap();
+        let mut buffer = Vec::with_capacity(SNAP_CHUNK_LEN);
+        buffer.push(name.len() as u8);
+        buffer.extend_from_slice(name.as_bytes());
+        let mut f = File::open(&path)?;
+        let mut off = buffer.len();
+        loop {
+            unsafe { buffer.set_len(SNAP_CHUNK_LEN) }
+            let readed = f.read(&mut buffer[off..])?;
+            limiter.consume(readed);
+            let new_len = readed + off;
+            total_sent += new_len as u64;
+            unsafe {
+                buffer.set_len(new_len);
+            }
+            let mut chunk = SnapshotChunk::default();
+            chunk.set_data(buffer);
+            sender
+                .feed((chunk, WriteFlags::default().buffer_hint(true)))
+                .await?;
+            if readed == 0 {
```

Comment:
Wasteful to always send one empty chunk. Should loop for eof:
```rust
let mut readed = 0;
loop {
  let r = f.read(&mut buffer[off..])?;
  if r == 0 {
    break;
  } else {
    readed += r;
  }
}
```

Reply:
There's a problem without sending the empty chunk. In fact it's used as a separator when the file can be just sent with full chunk(s). Otherwise the receiver cannot know when to start a new file. @tabokie

Reply:
how about carry with the file size in the first chunk size?